modules/command-engine/javadsl/src/main/scala/surge/javadsl/command/CommandModels.scala

```diff
@@ -2,109 +2,77 @@
+        handleEvent(state.asJava, event).asScala.map { newState =>
+          ctx.updateState(newState.asScala).reply(state => state)
+        }(ExecutionContext.global)
+      }
     }
 }
 
-/**
- * Trait for implementing a rejectable command model. Unlike `AggregateCommandModel` this type of command model may choose to accept or reject commands.
- * @tparam Agg
- *   state type
- * @tparam Cmd
- *   command type
- * @tparam Rej
- *   rejection type
- * @tparam Evt
- *   event type
- */
-trait RejectableAggregateCommandModel[Agg, Cmd, Rej, Evt] extends AggregateCommandModelCoreTrait[Agg, Cmd, Rej, Evt] {
-
-  /**
-   * Process a command
-   * @param ctx
-   *   the surge context
-   * @param aggregate
-   *   the current aggregate state
-   * @param command
-   *   the command to process
-   * @return
-   *   a CompleteableFuture of Either a rejection or a sequence or events generated by processing the command
-   */
-  def processCommand(ctx: Context, aggregate: Optional[Agg], command: Cmd): CompletableFuture[Either[Rej, Seq[Evt]]]
+trait ContextAwareAggregateCommandModel[Agg, Cmd, Evt] extends SurgeProcessingModelCoreTrait[Agg, Cmd, Evt] {
+  def processCommand(ctx: Context[Agg, Evt], aggregate: Optional[Agg], command: Cmd): CompletableFuture[Context[Agg, Evt]]
+  def handleEvent(aggregate: Optional[Agg], event: Evt): Optional[Agg]
 
-  /**
-   * Handle an event
-   * @param ctx
-   *   the surge context
-   * @param aggregate
-   *   the current aggregate state
-   * @param event
-   *   the event to apply
-   * @return
-   *   the some resulting aggregate state or `None` to remove the state.
-   */
-  def handleEvent(ctx: Context, aggregate: Optional[Agg], event: Evt): Optional[Agg]
+  final def toCore: SurgeProcessingModel[Agg, Cmd, Evt] =
+    new SurgeProcessingModel[Agg, Cmd, Evt] {
+      override def handle(ctx: SurgeContext[Agg, Evt], state: Option[Agg], msg: Cmd)(implicit ec: ExecutionContext): Future[SurgeContext[Agg, Evt]] = {
+        processCommand(Context(ctx), state.asJava, msg).asScala.map(_.toCore)
+      }
 
-  final def toCore: CommandHandler[Agg, Cmd, Rej, Evt] =
-    new CommandHandler[Agg, Cmd, Rej, Evt] {
-      override def processCommand(ctx: persistence.Context, state: Option[Agg], cmd: Cmd): Future[CommandResult] =
-        FutureConverters.toScala(RejectableAggregateCommandModel.this.processCommand(Context(ctx), state.asJava, cmd))
-      override def apply(ctx: persistence.Context, state: Option[Agg], event: Evt): Option[Agg] = handleEvent(Context(ctx), state.asJava, event).asScala
+      override def applyAsync(ctx: SurgeContext[Agg, Evt], state: Option[Agg], event: Evt): Future[SurgeContext[Agg, Evt]] = {
+        val newState = handleEvent(state.asJava, event)
+        Future.successful(ctx.updateState(newState.asScala))
+      }
     }
 }
+
+// TODO Add explicit reply type model to prove out the abstraction
```

Comment:
Need to do this TODO still

Reply:
Also need to add back a model where the domain can explicitly reject commands somehow